### PR TITLE
Fix #8285: napoleon can use PEP526 annotations

### LINKED ISSUE #8285: [Napoleon] Retrieve type of attributes from type hints when using google-style
Following [google style guide for classes](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#384-classes), I added an `Attributes:` section my classes docstring. As I have [PEP 526](https://www.python.org/dev/peps/pep-0526/) annotations (because I use `attrs`), I did not add types in the docstring., as per the style guide.

When generating the documentation with `.. automodule::` from `autodoc`, with `napoleon`, the attributes are not typed. I tried using the `:undoc-members:` flag of `automodule`, but this resulted in duplicated attributes: one with the explanation, the other with the type-hint.

While it is possible to not use type hints with `attrs` and replace `attr.s(auto_attribs=True)` by `attr.ib()`, this is not an option for e.g. [`dataclasses`](https://docs.python.org/3/library/dataclasses.html). I also tried `napoleon_use_ivar=True`, which silenced sphinx's warnings but still resulted in two definition of the attribute in the documentation.

It would be nice if `napoleon` (or `autodoc`) extracted the type hint of the class, or merged the attributes when using `:undoc-members:`. That would avoid duplicated types definitions in either the code or the doc. Currently, either the code has an annotation and a type in the docstring, or the generated documentation has two entry for each attribute.

This might be related to #7582  and #4074 .

**Additional info**
- sphinx version `3.2.1`
- [`conf.py`](https://github.com/QuentinSoubeyran/pheres/blob/dev/docs/conf.py)
- [code being autodocumented](https://github.com/QuentinSoubeyran/pheres/blob/dev/src/pheres/exceptions.py)

### PR DESCRIPTION
**Pull request by QuentinSoubeyran**
### Feature
Add an option so that napoleon fetches PEP 526-style  attributes annotation  in google style

### Purpose
This simplifies the use of the google style with PEP 526. The google style recommends adding the type of attributes in the class docstring only if the attribute doesn't have a type annotation in the code. As some libraries (e.g. `dataclasses`, `attrs`) uses those annotations and makes them mandatory, this features avoids duplicated type annotations.

### Detail
- added a new option to `napoleon`, `napoleon_google_attr_annotations` (boolean, defaults to False)
- If the option is `True`, in google style, if an attribute doesn't have a type in the docstring, the class is inspected for PEP 526-style annotations for that attribute.
- documented the new option in `napoleon.rst`
- added an example in the `example_google.py` in the documentation
- added a test case in `test/test_ext_napoleon_docstring.py`

*Note:* Default behaviour is not changed, as `napoleon_google_attr_annotations` defaults to False

### Relates
- #8285 

### Notes
I could only test python 3.8 and 3.9rc1, as `tox` couldn't find other versions.
Some `lualatex` test fails on my machine (including on the`3.x` branch so I don't think it's due to this change):
 - `tests/test_build_latex.py::test_build_latex_doc[lualatex-howto]`
- `tests/test_build_latex.py::test_build_latex_doc[lualatex-manual]`

I'm new to sphinx, so If I forgot anything I'm happy to improve this PR ! I'm not sure about the option name in particular, perhaps a better name can be found.

### CODE REVIEW

**Comment by tk0miya** [comment]
Could you rebase this into 3.x branch? It seems you're trying to merge master branch (it is mainly for developing 4.x) to 3.x branch.

**Comment by QuentinSoubeyran** [comment]
Hi, it should be good now, sorry for the mess. I used two force push to undo my erroneous merge and for the rebase, I hope this is fine.

There are still limitations with the code, I think: for instance, it does not comply with autodoc's `autodoc_type_aliases` config. I would expect type hints loaded from type annotations to be handled exactly like those autodoc derives from function signatures. They are very close in python code and have the same semantic.
Perhaps part of this feature should be implemented in autodoc ? The feature lies at the interface between autodoc and napoleon, as part code part docstring.

I'm open to suggestions or pointers on how to do that properly

**Comment by QuentinSoubeyran** [comment]
When using my branch in my project (`pheres`), I'm getting weird `Handler <function warn_missing_reference at 0x...> for event 'warn-missing-reference' threw an exception (exception: 'NoneType' object has no attribute 'name')` errors, if there are missing references in my docstrings. It looks like the `domain` argument is None in [`domain/std.py`](https://github.com/QuentinSoubeyran/sphinx/blob/5e578d197865488696903b3e248a25ed1440d7a8/sphinx/domains/std.py#L1108).
I can't reproduce if I install sphinx using `pip install git+https://github.com/sphinx-doc/sphinx@3.x`.

I believe my `git rebase 3.x` went awry, because my branch reports running sphinx `v3.4.0+/fab8e7b`, while install from pip as above reports sphinx `v3.3.0+/fab8e7b`. I don't think I modified any code regarding missing reference or event handling.

I couldn't pinpoint if is a bug in `v3.4.0+/fab8e7b`, a bug in my code or a failed rebase. Any ideas ?

File: sphinx/ext/napoleon/docstring.py
Comment (tk0miya) [inline sphinx/ext/napoleon/docstring.py]: Now I posted https://github.com/sphinx-doc/sphinx/pull/8441 not to copy the function more.

File: sphinx/ext/napoleon/docstring.py
Comment (tk0miya) [inline sphinx/ext/napoleon/docstring.py]: Please pass `self._config.autodoc_type_aliases` as the 3rd argument as same as I did in https://github.com/sphinx-doc/sphinx/pull/8435. Then it also supports `autodoc_type_aliases`!

I guess `napoleon_type_aliases` is also useful for this case.
https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html

File: doc/usage/extensions/napoleon.rst
Comment (tk0miya) [inline doc/usage/extensions/napoleon.rst]: Now I'm thinking to apply this to numpy also. There is no reason to limit this feature to google style only.

File: doc/usage/extensions/napoleon.rst
Comment (tk0miya) [inline doc/usage/extensions/napoleon.rst]: Please add `versionadded:: 3.4` after this block :-)

File: sphinx/ext/napoleon/__init__.py
Comment (tk0miya) [inline sphinx/ext/napoleon/__init__.py]: Is there any reason to disable this feature by default?

File: sphinx/ext/napoleon/docstring.py
Comment (tk0miya) [inline sphinx/ext/napoleon/docstring.py]: I think we can use this feature to module-level variables also.

**Review by tk0miya — CHANGES_REQUESTED** [review]
Good work! I love this implementation. Let's make this better and merge into the next release!

**Comment by tk0miya** [comment]
AFAIK, the errors around `warn-missing-reference` were added into the 3.x branch accidentally, but they have been resolved now. Please let me know if you still see the error.

File: sphinx/ext/napoleon/docstring.py
Comment (QuentinSoubeyran) [reply sphinx/ext/napoleon/docstring.py]: I had to use `getattr()` as tests failed due to `AttributeError`s. Should I leave that there as a safety, or change the test to be properly configured (or both ?)

File: doc/usage/extensions/napoleon.rst
Comment (QuentinSoubeyran) [reply doc/usage/extensions/napoleon.rst]: I started looking at that, I'm not too sure how to do it yet. Is `sphinx.ext.napoleon.docstring.NumpyDocstring::_consume_field()` the right place ?

File: sphinx/ext/napoleon/__init__.py
Comment (QuentinSoubeyran) [reply sphinx/ext/napoleon/__init__.py]: I was waiting on some sort of approval of the feature before enabling it by default. It is now enabled

File: sphinx/ext/napoleon/docstring.py
Comment (QuentinSoubeyran) [reply sphinx/ext/napoleon/docstring.py]: What would be the value of `_what` for module-level variables ? Is it `module`, or another value not reflected in `GoogleDcostring` own docstring ?

**Comment by QuentinSoubeyran** [comment]
Thank you for your review and help !

Following your comment on the `warn-missing-reference`, I think I rebased on the local `3.x` branch instead of the remote, without pulling first (oops). I rebased on the upstream after a fetch to fix that (this required another force push).

I also started implementing the requested changes:
- the feature is now called `napoleon_attr_annotations`, to prepare for numpy style implementation
- the feature is enabled by default
- refactored code to use `sphinx.util.typing.get_type_hints()`, with support for `config.napoleon_type_aliases` and `autodoc_type_aliases`
- added a small fix in `sphinx.util.typing.get_type_hints()` so that the `globalns` argument isn't ignored and forced to `None`. This is minor so I thought it would be OK to do in this PR instead of a new one.
- refactored tests to prepare testing module-level variables and numpy style implementation
- added the version in which the feature will be added in the doc

What is left to do is:
- implement the feature for numpy style (I haven't figured out exactly where to do that yet, see requested change discussion above)
- implement the feature for module-level variables (see requested change discussion above)

EDIT: added mention of the `get_type_hints` fix

File: sphinx/ext/napoleon/docstring.py
Comment (QuentinSoubeyran) [reply sphinx/ext/napoleon/docstring.py]: Just understood what you meant, activated it for `module` too

File: sphinx/ext/napoleon/__init__.py
Comment (tk0miya) [inline sphinx/ext/napoleon/__init__.py]: True by default?

File: sphinx/ext/napoleon/docstring.py
Comment (tk0miya) [inline sphinx/ext/napoleon/docstring.py]: I feel this must be a bug of testing framework of napoleon. I feel this is ugly. But it's not related to this change. So let's fix in another PR.

File: sphinx/util/typing.py
Comment (tk0miya) [inline sphinx/util/typing.py:69]: Oops...

**Review by tk0miya — APPROVED** [review]
LGTM with nits.

File: sphinx/ext/napoleon/__init__.py
Comment (QuentinSoubeyran) [reply sphinx/ext/napoleon/__init__.py]: Oops, thanks for catching that ! Fixed it

File: sphinx/ext/napoleon/docstring.py
Comment (QuentinSoubeyran) [reply sphinx/ext/napoleon/docstring.py]: Works for me. I believe this is because in napoleon test cases, the object passed as a`config` is an instance of napoleon's own `Config` class, which of course lacks autodoc's config options

Comment (QuentinSoubeyran) [reply sphinx/ext/napoleon/docstring.py]: Actually,  the `or {}` part is unnecessary. I assumed an unspecified option could be `None`, but the doc saying the default is `{}`. I removed `or {}` and tests and my own project using this feature work correctly.

**Comment by QuentinSoubeyran** [comment]
@tk0miya Fixed the documentation error and improved a bit the ugly code.
FYI, I still have the `warn-missing-reference` issue in my project with this branch, but also with the current `3.x` branch (commit 5a0e1231ad3fce9405ea52d90b21153e4ea86a27). If that helps, this is caused by the config line `default_role = "any"`. I'll do a proper bug report for that

EDIT: see #8491

**Comment by tk0miya** [comment]
Merged. Thank you for your great work!

**Comment by QuentinSoubeyran** [comment]
My pleasure !

**Pull request merged**
